test/hotreload/vshard_test.lua

```diff
@@ -0,0 +1,168 @@
+local fio = require('fio')
+local log = require('log')
+local fiber = require('fiber')
+local errors = require('errors')
+local t = require('luatest')
+local g = t.group()
+
+local helpers = require('test.helper')
+
+local function reload(srv)
+    local ok, err = srv.net_box:eval([[
+        return require("cartridge.roles").reload()
+    ]])
+
+    t.assert_equals({ok, err}, {true, nil})
+end
+
+g.before_all(function()
+    g.cluster = helpers.Cluster:new({
+        datadir = fio.tempdir(),
+        use_vshard = true,
+        server_command = helpers.entrypoint('srv_basic'),
+        cookie = helpers.random_cookie(),
+        replicasets = {
+            {alias = 'R', roles = {'vshard-router'}, servers = 1},
+            {alias = 'SA', roles = {'vshard-storage'}, servers = 2},
+            {alias = 'SB', roles = {'vshard-storage'}, servers = 2}
+        },
+    })
+    g.cluster:start()
+
+    local test_schema = {
+        engine = 'memtx',
+        is_local = false,
+        temporary = false,
+        format = {
+            {name = 'bucket_id', type = 'unsigned', is_nullable = false},
+            {name = 'record_id', type = 'unsigned', is_nullable = false},
+        },
+        indexes = {{
+            name = 'pk', type = 'TREE', unique = true,
+            parts = {{path = 'record_id', is_nullable = false, type = 'unsigned'}},
+        },  {
+            name = 'bucket_id', type = 'TREE', unique = false,
+            parts = {{path = 'bucket_id', is_nullable = false, type = 'unsigned'}},
+        }},
+        sharding_key = {'record_id'},
+    }
+    g.cluster.main_server.net_box:call('cartridge_set_schema',
+        {require('yaml').encode({spaces = {test = test_schema}})}
+    )
+
+    g.R1 = assert(g.cluster:server('R-1'))
+    g.SA1 = assert(g.cluster:server('SA-1'))
+    g.SA2 = assert(g.cluster:server('SA-2'))
+
+    g.insertions_passed = {}
+    g.insertions_failed = {}
+end)
+
+g.after_all(function()
+    g.cluster:stop()
+    fio.rmtree(g.cluster.datadir)
+end)
+
+local function _insert(cnt, label)
+    local ret, err = g.R1.net_box:eval([[
+        local ret, err = package.loaded.vshard.router.callrw(...)
+        if ret == nil then
+            return nil, tostring(err)
+        end
+        return ret
+    ]], {1, 'box.space.test:insert', {{1, cnt, label}}})
+
+    if ret == nil then
+        log.error('CNT %d: %s', cnt, err)
+        table.insert(g.insertions_failed, {cnt = cnt, err = err})
+    else
+        table.insert(g.insertions_passed, ret)
+    end
+    return true
+end
+
+local highload_cnt = 0
+local function highload_loop(label)
```

Comment:
Do you have something like that for http route?

Reply:
If you're asking for a useful snippet, no, I don't. And if your question is about test coverage, then it doesn't matter. Both HTTP and net.box calls are canceled similarly on reload.